### PR TITLE
[BD-21] v4.0.0: remove deprecated waffle API

### PR DESCRIPTION
**Pull request by regisb**
**Description:** remove deprecated waffle API 

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation

**Monitoring shows no results for:**

@regisb: You can cull or change this last as necessary.  I am just noting what I am seeing under monitoring.

**This PR:**
- [x] deprecated_incompatible_legacy_waffle_class
- [x] edx_toggles.toggles.internal.waffle.legacy
- [x] deprecated_waffle_legacy_method
- [x] deprecated_waffle_method
- [x] deprecated_module_not_supplied

**Requires separate edx-platform PR:**
- [ ] deprecated_waffle_utils

**Merge deadline:** Feb 24 2021

**Reviewers:**
- [ ] @robrap 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

This is a WIP PR which we use to list all the things to be done in v3.0.0:

- Remove `toggles.__future__` module
- Some monitoring?

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-700](https://openedx.atlassian.net/browse/BLENDED-700) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by regisb** [comment]
@robrap Could you please list the attributes that you keep seeing in monitoring? What I would like to do is to get rid of all calls to `set_custom_attribute` in edx-toggles (after making sure that there is no deprecated usage left behind).

**Comment by robrap** [comment]
@regisb: 
1. I had added to the PR description everything that I think can be removed because monitoring is not showing any usage. I do want you to ensure that each makes sense.
2. Toggles 2.1.0 has not made it to edx-platform yet as of my writing this (see https://github.com/edx/edx-platform/blob/master/requirements/edx/base.txt#L115). Once the upgrade has been deployed, your new deprecated custom attribute should appear for me in monitoring.  Then, I'll merge https://github.com/edx/edx-platform/pull/26040 and let you know when that custom attribute is no longer appearing.
3. The other major custom attribute is `deprecated_compatible_legacy_waffle_class`, but this shows all the Legacy classes which need to be fixed before we remove in 4.0.0.

Is there anything else you are looking for?

**Comment by regisb** [comment]
This should be ready for review @robrap. I hope I didn't misunderstand the removal of monitoring calls.

File: edx_toggles/toggles/__init__.py
Comment (robrap) [inline edx_toggles/toggles/__init__.py:17]: 1. Probably because it is late, it took me a while to understand what was going on here.  I think you are renaming `deprecated_compatible_legacy_waffle_class` to `deprecated_legacy_waffle_class`, which I will need to change in my monitoring.  Is that correct?  Please update the changelog.  Thanks.
2. Could we simply rename `WaffleSwitch` to `LegacyWaffleSwitch`, as well as the others, inside `.internal.waffle.legacy`? This would make it more obvious everywhere, including in the custom attributes, that these are the Legacy classes.  That is what having these class names in this file had provided.

**Review by robrap — COMMENTED** [review]
@regisb: Thanks for these changs. I know it took me a while to get to it. Sorry about the conflicts.

Note: in the PR description, there is another deprecated custom attribute that would requiree a separate PR in edx-platform.

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/flag.py:86]: > 1. This custom attribute is somewhat independent of all the deprecated custom attributes, but it also probably isn't necessary, so feel free to drop it. Note: We should add a comment to the changelog for each custom attribute that we are removing.

:ok_hand: 

> 2. Unrelated, except by proximity, I wish we had killed log_prefix that is also in this file. Can we do so while we are making backward incompatible changes? Note: if we do this, just don't mention that it is "unused" in th changelog, unless you check on this.

Actually, the `log_prefix` argument is used in quite a few places in edx-platform. Should we simply remove it there?

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:20]: Given that this PR contains little else, I would be inclined to remote `__future__`. But we should probably remove all usage in edx-ora2, edx-val, completion and edx-platform before merging this.

File: edx_toggles/toggles/__init__.py
Comment (regisb) [reply edx_toggles/toggles/__init__.py:17]: > Probably because it is late, it took me a while to understand what was going on here. I think you are renaming deprecated_compatible_legacy_waffle_class to deprecated_legacy_waffle_class, which I will need to change in my monitoring. Is that correct? Please update the changelog. Thanks.

I believe this is correct -- but I'm still very fuzzy about custom attributes and how they are used on your side, so I'm not sure what is the final impact.

> Could we simply rename WaffleSwitch to LegacyWaffleSwitch, as well as the others, inside .internal.waffle.legacy?

Yes. I'll do that.

**Comment by regisb** [comment]
@robrap I rebased my changes on top of master to resolve conflicts. Note that the target version is now v4.0.0, because v3.1.0 was released earlier.

File: CHANGELOG.rst
Comment (robrap) [reply CHANGELOG.rst:20]: Also, if you end up having higher priorities, the fix is pretty simple and we could let others take care of it as well.  I'll leave it to you to decide if you want to open 4 PRs for this.

File: edx_toggles/toggles/internal/waffle/flag.py
Comment (robrap) [reply edx_toggles/toggles/internal/waffle/flag.py:86]: Sorry.  Let's forget about `log_prefix` for now.  I've documented this, but it is a lower priority.  It just seems not so useful, so it bugs me.

File: edx_toggles/toggles/__init__.py
Comment (robrap) [reply edx_toggles/toggles/__init__.py:17]: Great.  To clarify, [in this code](https://github.com/edx/edx-toggles/blob/e654795315bf05a6053534da5b10f672cffb55d0/edx_toggles/toggles/internal/waffle/legacy.py#L58-L63), we refer to the class name in the value, and it would be nice if it contain `Legacy`.  The requested change will make that happen.

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:20]: Nit: "now unused" => "now unnecessary"

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:183]: Oops.  I should have been more clear.  Since `deprecated_waffle_legacy_method` is showing me no usage, I think we are ready to delete this whole property.  Same is true for all the other things I noted are not appearing in monitoring.  This clean-up will make it so that the only difference between legacy and non-legacy will be the namespace in the constructor, ensuring the remaining refactors to remove legacy are extremely simple.

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:20]: I'll take care of the PRs.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (regisb) [reply edx_toggles/toggles/internal/waffle/legacy.py]: Right! This explains why you were explicitly listing custom attributes to be removed. I'll remove the methods and add them to the changelog.

**Review by robrap — APPROVED** [review]
Thank you! Let’s wait for future proofing to land before merging, and all upgrades to land in edx-platform. Are you planning on doing the upgrade, or are we going to wait for the automated jobs?

**Comment by regisb** [comment]
> Are you planning on doing the upgrade, or are we going to wait for the automated jobs?

I can create a dummy PR which will upgrade edx-toggles, edx-val, completion, edx-ora2. It tests pass then it means we can merge this and let the auto-updated do its magic.

File: CHANGELOG.rst
Comment (robrap) [reply CHANGELOG.rst]: @regisb: Thanks for the PRs.  FYI: I think you missed this one: https://github.com/edx/edx-ora2/blob/master/openassessment/xblock/config_mixin.py#L7

**Pull request merged**
